src/lib/es2015.generator.d.ts

```diff
@@ -1,4 +1,12 @@
-interface Generator extends Iterator<any> { }
+/// <reference lib="es2015.iterable" />
+
+interface Generator<T = unknown, TReturn = any, TNext = unknown> extends Iterator<T, TReturn, TNext> {
+    // NOTE: 'next' is defined using a tuple to ensure we report the correct assignability errors in all places.
+    next(...args: [] | [TNext]): IteratorResult<T, TReturn>;

```

Comment:
isn't `...args: [] | [TNext]` effectively `arg?: TNext`?
Does it mean that allows to not pass argument even if `TNext` is not undefined?
Could a conditional type work here: `...args: undefined extends TNext ? [TNext?] : [TNext]`?

Reply:
The first call to `next()` starts the generator and runs it until the first `yield`. Therefore, the argument passed to this first call is completely ignored.

There's no good way to describe this statically. You'd need to know whether the code has already called `next()` to figure out what the signature should be. This is intractable, so instead this type definition describes both the first call and subsequent calls.

Reply:
`next(...args: [] | [TNext])` is very similar to an overload list of `next(); next(value: TNext);`, however TypeScript compares them differently. The problem is that when we compare the methods of two types, we check whether *any* overload in the source is assignable to *any* overload in the target. Unfortunately, that means `next()` in the source is assignable to `next()` in the target, even if the source and the target really shouldn't be assignable.

Conditional types do not work here because conditional types have even stricter assignability rules that would prevent anyone from easily subtyping `Iterator`.